tasmota/xsns_96_bl6523.ino

```diff
@@ -0,0 +1,353 @@
+    } 
+  }
+}
+
+/*********************************************************************************************/
+
+void Bl6523Init(void)
+{
+  Bl6523.type = 0;
+  if ((PinUsed(GPIO_BL6523_RX)) && (PinUsed(GPIO_BL6523_TX)))
+  {
+    Bl6523RxSerial = new TasmotaSerial(Pin(GPIO_BL6523_RX), -1, 1);
+    Bl6523TxSerial = new TasmotaSerial(Pin(GPIO_BL6523_TX), -1, 1);
+    if ((Bl6523RxSerial->begin(BL6523_BAUD)) && (Bl6523TxSerial->begin(BL6523_BAUD)))
+    {
+      if (Bl6523RxSerial->hardwareSerial())
+      {
+        ClaimSerial();
+      }
+      if (Bl6523TxSerial->hardwareSerial())
+      {
+        ClaimSerial();
+      }
+      Bl6523.type = 1;
+      AddLog(LOG_LEVEL_DEBUG, PSTR("BL6523 Init Success " ));
+    }
+  }
+}
+
+#ifdef USE_WEBSERVER
+const char HTTP_BL6523_SNM[] PROGMEM = "{s}BL6523 Smart Energy Monitor{m}{e}"; // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+const char HTTP_BL6523_SNS[] PROGMEM = "{s}       %s  {m}%s {e}"; // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+#endif  // USE_WEBSERVER
+
+void Bl6523Show(bool json)
+{
+  uint32_t powf_word = 0;
+  float amps = 0.0f, volts = 0.0f, freq = 0.0f, watts = 0.0f, powf = 0.0f, watthr = 0 .0f;
+  char amps_str[12], volts_str[12], freq_str[12];
+  char watts_str[12], powf_str[12], watthr_str[12];
+
+  if (Bl6523.valid)
+  {
+
+   amps = (float)Bl6523.amps / BL6523_DIV_AMPS;
+   volts =  (float)Bl6523.volts / BL6523_DIV_VOLTS;
+   freq = (float)Bl6523.freq / BL6523_DIV_FREQ;
+   watts = (float)Bl6523.watts / BL6523_DIV_WATTS;
+
+    /* Power factor =(sign bit)*((PF[22]×2^－1）＋（PF[21]×2^－2）＋。。。)
+       Eg., reg value  0x7FFFFF(HEX) -> PF 1, 0x800000(HEX) -> -1, 0x400000(HEX) -> 0.5
+    */
+   powf = 0.0f;
+   powf_word = Bl6523.powf & 0x7fffff; //Extract the 23 bits
+   for (int i = 0; i < 23; i++){ // Accumulate powf from 23 bits
+    powf += ((powf_word >> (22-i)) * pow(2,(0-(i+1)))); 
+    powf_word = powf_word & (0x7fffff >> (1+i));
+   }
+   powf = (Bl6523.powf >> 23) ? (0.0f - (~powf)) : powf; // Negate if sign bit(24) is set
+   
+   watthr = (float)Bl6523.watthr / BL6523_DIV_WATTHR;
+
+   ext_snprintf_P(amps_str, sizeof(amps_str), PSTR("%3_f"), &amps);
+   ext_snprintf_P(volts_str, sizeof(volts_str), PSTR("%2_f"), &volts);
+   ext_snprintf_P(freq_str, sizeof(freq_str), PSTR("%2_f"), &freq);
+   ext_snprintf_P(watts_str, sizeof(watts_str), PSTR("%3_f"), &watts);
+   ext_snprintf_P(powf_str, sizeof(powf_str), PSTR("%2_f"), &powf);
+   ext_snprintf_P(watthr_str, sizeof(watthr_str), PSTR("%3_f"), &watthr);
+
+    if (json)
+    {
+      ResponseAppend_P(PSTR(",\"BL6523\":{"));
+      ResponseAppend_P(PSTR("\"Amps\":%s,"), amps_str);
+      ResponseAppend_P(PSTR("\"Volts\":%s,"), volts_str);
+      ResponseAppend_P(PSTR("\"Freq\":%s,"), freq_str);
+      ResponseAppend_P(PSTR("\"Watts\":%s,"), watts_str);
+      ResponseAppend_P(PSTR("\"Powf\":%s,"), powf_str);
+      ResponseAppend_P(PSTR("\"WattHr\":%s"), watthr_str);
+      ResponseJsonEnd();
```

Comment:
Don't you your own name strings for reporting the measurments.
There are standard Tasmota name available in `i18n.h` such as D_JSON_CURRENT, D_JSON_VOLTAGE, D_JSON_FREQUENCY, D_JSON_POWERUSAGE, D_JSON_POWERFACTOR, D_JSON_ENERGY

Moreover, shouldn't this driver be a XNRG driver instead of XSNS and use the standard Tasmota ENERGY reporting model ?